components/sceneutil/recastmesh.cpp

```diff
@@ -0,0 +1,60 @@
+#include "navmesh.hpp"
+#include "detourdebugdraw.hpp"
+
+#include <components/detournavigator/settings.hpp>
+#include <components/detournavigator/recastmesh.hpp>
+
+#include <RecastDebugDraw.h>
+
+#include <osg/Group>
+
+namespace
+{
+    std::vector<float> calculateNormals(const DetourNavigator::RecastMesh& recastMesh)
+    {
+        // Adapted copy-paste from RecastDemo
+        std::vector<float> result(recastMesh.getTrianglesCount() * 3);
+        const auto& vertices = recastMesh.getVertices();
+        const auto& indices = recastMesh.getIndices();
+        for (std::size_t i = 0, n = recastMesh.getTrianglesCount() * 3; i < n; i += 3)
+        {
+            const float* v0 = &vertices[indices[i] * 3];
+            const float* v1 = &vertices[indices[i + 1] * 3];
+            const float* v2 = &vertices[indices[i + 2] * 3];
+            float e0[3], e1[3];
+            for (int j = 0; j < 3; ++j)
+            {
+                e0[j] = v1[j] - v0[j];
+                e1[j] = v2[j] - v0[j];
+            }
+            float* normal = result.data() + i;
+            normal[0] = e0[1] * e1[2] - e0[2] * e1[1];
+            normal[1] = e0[2] * e1[0] - e0[0] * e1[2];
+            normal[2] = e0[0] * e1[1] - e0[1] * e1[0];
+            float d = std::sqrt(normal[0] * normal[0] + normal[1] * normal[1] + normal[2] * normal[2]);
+            if (d > 0)
+            {
+                d = 1.0f / d;
+                normal[0] *= d;
+                normal[1] *= d;
+                normal[2] *= d;
+            }
+        }
+        return result;
+    }
+}
```

Comment:
This would probably be much easier to read if it used OSG vector types instead of floats.

Reply:
Okay, I see that this gets passed to some external code from Recast. I *think* there's a nice, clean, cross-platform way of getting an OSG vector array or buffer or something as contiguous float memory (as otherwise there'd be no way to put it on a GPU), but if it turns out that accessing it is super-ugly, I can excuse doing vector operations on scalar types.

Reply:
I had some concerns about using `osg::Vec3f` but it seems  there is no difference in performance: http://quick-bench.com/sssJs5iB8J_GkZlmq6X10iHdIK0
Also quite interesting see here clang 8 beating gcc 9.1: http://quick-bench.com/6ElyfweLhiS85UxO5AwN6gT09_Y

Reply:
There's lots of inlining, so there's no practical difference once that's been applied.